src/generic/stage2/numberparsing.h

```diff
@@ -361,32 +394,19 @@ really_inline bool parse_number(UNUSED const uint8_t *const src,
   if (found_minus) {
     ++p;
     negative = true;
-    if (!is_integer(*p)) { // a negative sign must be followed by an integer
-#ifdef JSON_TEST_NUMBERS // for unit testing
-      found_invalid_number(src);
-#endif
-      return false;
-    }
+    // a negative sign must be followed by an integer
+    // TODO this is a redundant check
```

Comment:
Given that you have identified that it is redundant, why not remove it?

Reply:
Because when I tried, it got slower :)

Reply:
Here's what happens when you remove the redundant check:

| File                                | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                                 |    --- |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| citm_catalog                        |  26987 |  34.1 |  34.1 |    0% | 130.7 | 130.7 |    0% |    164 |    179 |   -9% |      2 |      7 | -250% |  76725 |  71783 |    6% |
| github_events                       |   1017 |  34.3 |  34.4 |    0% | 129.8 | 129.8 |    0% |      4 |      4 |    0% |      0 |      0 |    0% |   1849 |   1164 |   37% |
| mesh.pretty                         |  24646 |  35.7 |  35.8 |    0% | 137.7 | 137.7 |    0% |     39 |     42 |   -7% |      0 |      0 |    0% |  74481 |  69348 |    6% |
| random                              |   7976 |  57.6 |  57.8 |    0% | 209.3 | 209.3 |    0% |   1013 |   1005 |    0% |      0 |      0 |    0% |  29778 |  27763 |    6% |
| marine_ik                           |  46616 |  53.8 |  54.1 |    0% | 180.9 | 180.9 |    0% |   8882 |   8771 |    1% |  72327 |  74782 |   -3% | 191265 | 186677 |    2% |
| canada                              |  35172 |  45.0 |  45.3 |    0% | 157.6 | 157.6 |    0% |   5518 |   5354 |    2% |   7680 |  10541 |  -37% | 122181 | 115501 |    5% |
| apache_builds                       |   1988 |  36.7 |  37.0 |    0% | 139.3 | 139.3 |    0% |     21 |     31 |  -47% |      0 |      0 |    0% |   5757 |   5118 |   11% |
| twitter                             |   9867 |  43.1 |  43.5 |    0% | 148.8 | 148.8 |    0% |   1721 |   1752 |   -1% |      0 |      0 |    0% |  29042 |  27124 |    6% |
| mesh                                |  11306 |  45.9 |  46.4 |   -1% | 174.2 | 174.2 |    0% |    725 |    884 |  -21% |      0 |      0 |    0% |  45939 |  43025 |    6% |
| gsoc-2018                           |  51997 |  35.7 |  36.4 |   -2% | 101.1 | 101.1 |    0% |  14683 |  14877 |   -1% |  18225 |  18775 |   -3% | 122163 | 116368 |    4% |
| update-center                       |   8330 |  43.4 |  44.9 |   -3% | 142.1 | 142.1 |    0% |   2541 |   2810 |  -10% |      0 |      0 |    0% |  26901 |  25241 |    6% |
| instruments                         |   3442 |  37.8 |  39.2 |   -3% | 143.1 | 143.1 |    0% |     95 |    300 | -215% |      0 |      0 |    0% |  11239 |  10321 |    8% |
| numbers                             |   2345 |  38.8 |  40.3 |   -4% | 144.3 | 144.3 |    0% |    125 |    275 | -120% |      0 |      0 |    0% |   7197 |   6988 |    2% |
| twitterescaped                      |   8787 |  41.1 |  42.9 |   -4% | 138.1 | 138.1 |    0% |   1882 |   2293 |  -21% |      0 |      0 |    0% |  26811 |  25046 |    6% |

Reply:
On the theory that the redundant check gives the compiler information it didn't otherwise have, I replaced the if() with SIMDJSON_ASSUME(is_integer(*p)). (The second if, not the one marked redundant here--it's not actually the redundant one.) That causes us to do `if (!is_integer(*p)) { __builtin_unreachable(); }`, which will tell the compiler that is_integer(*p) would succeed, but removes the actual call.

gcc 9, skylake, btw.

The result is exactly the same as when you remove the if(): things got slower.

So, I'm going to leave yet another big honking comment explaining it so people don't get rid of it unthinkingly. :(